src/Parser/Priest/Discipline/Modules/Spells/Grace.js

```diff
@@ -24,54 +25,100 @@ class grace extends Analyzer {
     statTracker: StatTracker,
   };
 
+  applyAbsorbEvents = [];
+
   graceHealing = 0;
-  baseHealing = 0; // Healing that mastery stemmed from
+
+  healingUnaffectedByMastery = 0;
+  healingUnbuffedByMastery = 0;
+  healingBuffedByMastery = 0;
+  atonement = 0;
 
   getGraceHealing(event) {
-    // Get our mastery level at the moment, this is required for temporary buffs
     const currentMastery = this.statTracker.currentMasteryPercentage;
     const masteryContribution = calculateEffectiveHealing(
       event,
       currentMastery
     );
-
     return masteryContribution;
   }
 
-  on_byPlayer_heal(event) {
-    // Check if the heal is whitelisted
-    if (!PRIEST_WHITELIST.includes(event.ability.guid)) {
+  on_byPlayer_absorbed(event){
+    const spellId = event.ability.guid;
+
+    if(!PRIEST_WHITELIST.includes(spellId)){
+      this.healingUnaffectedByMastery += event.amount;
       return;
     }
 
-    // Get the target
     const target = this.combatants.getEntity(event);
-
     if (!target) return;
 
-    // Mastery only buffs players benefitting from Atonement
-    if (!target.hasBuff(SPELLS.ATONEMENT_BUFF.id)) {
+    if(this.absorbApplicationWasMasteryBuffed(event)){
+      this.graceHealing += this.getGraceHealing(event);
+      this.healingBuffedByMastery += event.amount;
+    } else {
+      this.healingUnbuffedByMastery += event.amount;
+    }
+  }
+
+  // Finds the absorb buff application event to verify if the initial event was mastery buffed
+  absorbApplicationWasMasteryBuffed(event){
+    let i;
+    for (i = this.applyAbsorbEvents.length - 1; i >= 0; i--) {
```

Comment:
Instead of using a for loop, you can use `array.slice().reverse().find()` for this.